packages/firestore/src/api/database.ts

```diff
@@ -1170,590 +1125,119 @@ export class QueryDocumentSnapshot<T = PublicDocumentData>
-        `string or a DocumentReference, but it was: ` +
-        `${valueDescription(documentIdValue)}.`
-    );
-  }
-}
-
-/**
- * Validates that the value passed into a disjunctive filter satisfies all
- * array requirements.
- */
-function validateDisjunctiveFilterElements(
-  value: unknown,
-  operator: Operator
-): void {
-  if (!Array.isArray(value) || value.length === 0) {
-    throw new FirestoreError(
-      Code.INVALID_ARGUMENT,
-      'Invalid Query. A non-empty array is required for ' +
-        `'${operator.toString()}' filters.`
-    );
-  }
-  if (value.length > 10) {
-    throw new FirestoreError(
-      Code.INVALID_ARGUMENT,
-      `Invalid Query. '${operator.toString()}' filters support a ` +
-        'maximum of 10 elements in the value array.'
-    );
-  }
-}
-
-/**
- * Given an operator, returns the set of operators that cannot be used with it.
- *
- * Operators in a query must adhere to the following set of rules:
- * 1. Only one array operator is allowed.
- * 2. Only one disjunctive operator is allowed.
- * 3. NOT_EQUAL cannot be used with another NOT_EQUAL operator.
- * 4. NOT_IN cannot be used with array, disjunctive, or NOT_EQUAL operators.
- *
- * Array operators: ARRAY_CONTAINS, ARRAY_CONTAINS_ANY
- * Disjunctive operators: IN, ARRAY_CONTAINS_ANY, NOT_IN
- */
-function conflictingOps(op: Operator): Operator[] {
-  switch (op) {
-    case Operator.NOT_EQUAL:
-      return [Operator.NOT_EQUAL, Operator.NOT_IN];
-    case Operator.ARRAY_CONTAINS:
-      return [
-        Operator.ARRAY_CONTAINS,
-        Operator.ARRAY_CONTAINS_ANY,
-        Operator.NOT_IN
-      ];
-    case Operator.IN:
-      return [Operator.ARRAY_CONTAINS_ANY, Operator.IN, Operator.NOT_IN];
-    case Operator.ARRAY_CONTAINS_ANY:
-      return [
-        Operator.ARRAY_CONTAINS,
-        Operator.ARRAY_CONTAINS_ANY,
-        Operator.IN,
-        Operator.NOT_IN
-      ];
-    case Operator.NOT_IN:
-      return [
-        Operator.ARRAY_CONTAINS,
-        Operator.ARRAY_CONTAINS_ANY,
-        Operator.IN,
-        Operator.NOT_IN,
-        Operator.NOT_EQUAL
-      ];
-    default:
-      return [];
   }
-}
-
-function validateNewFilter(query: InternalQuery, filter: Filter): void {
-  debugAssert(filter instanceof FieldFilter, 'Only FieldFilters are supported');
 
-  if (filter.isInequality()) {
-    const existingField = getInequalityFilterField(query);
-    if (existingField !== null && !existingField.isEqual(filter.field)) {
-      throw new FirestoreError(
-        Code.INVALID_ARGUMENT,
-        'Invalid query. All where filters with an inequality' +
-          ' (<, <=, >, or >=) must be on the same field. But you have' +
-          ` inequality filters on '${existingField.toString()}'` +
-          ` and '${filter.field.toString()}'`
+  orderBy(
+    fieldPath: string | FieldPath,
+    directionStr?: PublicOrderByDirection
+  ): Query<T> {
+    try {
+      // The "as string" cast is a little bit of a hack. `orderBy` accepts the
```

Comment:
I'm surprised that we can't just make this different in our public typings vs are internal actual types. It seems like we could declare this in the .ts file and just omit it from the .d.ts?

Reply:
We are moving away from manually crafted typings. I still need to iron out some final details in https://github.com/firebase/firebase-js-sdk/pull/3790, but once that is merged, we will generate typings from source. All other firebase-exp SDKs are doing the same.